packages/langium/src/documents/document.ts

```diff
@@ -10,15 +10,57 @@ import { TextDocuments } from 'vscode-languageserver/node';
 import { AstNode } from '../syntax-tree';
 import { LangiumParser, ParseResult } from '../parser/langium-parser';
 import { AstNodeDescription } from '../references/scope';
-import { LangiumServices, LanguageMetaData, Stream, stream } from '..';
 import { URI } from 'vscode-uri';
+import { Mutable } from '../utils/ast-util';
+import { LanguageMetaData } from '../grammar/language-meta-data';
+import { LangiumServices } from '../services';
+import { stream, Stream } from '../utils/stream';
+
+export enum DocumentState {
+    Changed = 0,
+    Parsed = 1,
+    Processed = 2,
+    Indexed = 3,
+    Validated = 4
+}
+
+// = 'changed' | 'parsed' | 'processed' | 'indexed' | 'validated';
 
 export interface LangiumDocument {
     parseResult: ParseResult
     precomputedScopes?: PrecomputedScopes
-    valid: boolean
+    state: DocumentState
     textDocument: TextDocument
     uri: URI
+    reset(): void
+    onReset(listener: () => void): void
+}
+
+export class LangiumDocumentImpl implements LangiumDocument {
```

Comment:
AFAICT the only reason to change to a class is the local handling of "reset listeners", but the only consumer is the linker with its created Reference objects. I'd propose to keep the simple interface and implement a more direct solution:
 * Add a `references: Reference[]` property to LangiumDocument.
 * The linker adds every new Reference to that array.
 * When the DocumentBuilder resets an affected document, it asks the Linker to unlink the contained references and then clears the array.
WDYT?